test/jdk/javax/swing/JButton/bug4323121.java

```diff
@@ -30,50 +30,80 @@
  * @run main bug4323121
  */
 
-import java.awt.Graphics;
 import java.awt.Point;
 import java.awt.Robot;
+import java.awt.event.MouseAdapter;
 import java.awt.event.MouseEvent;
-import java.awt.event.MouseListener;
-import java.awt.event.MouseMotionListener;
+import java.awt.event.WindowAdapter;
+import java.awt.event.WindowEvent;
+import java.util.concurrent.CountDownLatch;
+
 import javax.swing.JButton;
 import javax.swing.JFrame;
 import javax.swing.SwingUtilities;
 
-public class bug4323121 {
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public final class bug4323121 extends MouseAdapter {
 
     static JFrame frame;
-    static testButton button;
-    static volatile Point pt;
-    static volatile int buttonW;
-    static volatile int buttonH;
-    static volatile boolean failed = false;
+    static JButton button;
+
+    static volatile Point buttonCenter;
+
+    private static final CountDownLatch windowGainedFocus = new CountDownLatch(1);
+
+    private static final CountDownLatch mouseEntered = new CountDownLatch(1);
+
+    // Thread-safe by using the mouseEntered latch
+    private static boolean modelArmed;
 
     public static void main(String[] args) throws Exception {
         Robot robot = new Robot();
         robot.setAutoDelay(100);
+
+        final bug4323121 eventHandler = new bug4323121();
         try {
             SwingUtilities.invokeAndWait(() -> {
+                button = new TestButton("gotcha");
+                button.addMouseMotionListener(eventHandler);
+                button.addMouseListener(eventHandler);
```

Comment:
You may get rid of `eventHandler` object (provided MouseMotionListener is not required) by adding the mouse listener similar to WindowFocusListener where you used WindowAdapter class and override WindowGainedFocus method. 
 
            button.addMouseListener(new MouseAdapter() {
                    @Override
                    public void mouseEntered(MouseEvent e) {
                        if (button.getModel().isArmed()) {
                            modelArmed = true;
                        }
                        mouseEntered.countDown();
                    }
                });

Reply:
Yes, it's an option.

I used the same approach in #21474 where the test object serves as an event handler. This avoids anonymous classes which add indentation.

Reply:
Yeah, this is an option. As I mentioned earlier that MouseMotionListener is not needed, so creating an eventHandler object may not be of great use. I agree if more than one listener needs to be added, creating an eventHandler object is a good idea.

Reply:
I removed the unneeded listener and inlined creation of the event handler.

Thus, I kept the current structure of test code. I prefer it this way for consistency with `JButton/bug4490179.java` in https://github.com/openjdk/jdk/pull/21474. Both tests are really similar, and I think consistent coding helps.

Reply:
Ok...

Reply:
If anyone else thinks using an anonymous class here makes the code shorter and cleaner, I'll go for it.

Reply:
Both approaches sounds okay, if had to pick one then adding an anonymous class similar to WindowAdapter() may be clearer

Reply:
Since you both for inlining the listener, I've amended the code.

Reply:
Looks clearer now.